halo2_proofs/src/plonk/circuit.rs

```diff
@@ -1545,6 +1685,472 @@ impl<F: Field> Gate<F> {
+        };
+        let mut witness = WitnessCollection {
+            k: self.k,
+            current_phase,
+            advice: vec![vec![Assigned::Zero; self.n]; self.cs.num_advice_columns],
+            instances: self.instances,
+            challenges,
+            // The prover will not be allowed to assign values to advice
+            // cells that exist within inactive rows, which include some
+            // number of blinding factors and an extra row for use in the
+            // permutation argument.
+            usable_rows: ..self.unusable_rows_start,
+            _marker: std::marker::PhantomData,
+        };
+
+        // Synthesize the circuit to obtain the witness and other information.
+        ConcreteCircuit::FloorPlanner::synthesize(
+            &mut witness,
+            self.circuit,
+            self.config.clone(),
+            self.cs.constants.clone(),
+        )
+        .expect("todo");
+
+        let column_indices = self
+            .cs
+            .advice_column_phase
+            .iter()
+            .enumerate()
+            .filter_map(|(column_index, phase)| {
+                if current_phase == *phase {
+                    Some(column_index)
+                } else {
+                    None
+                }
+            })
+            .collect::<BTreeSet<_>>();
+
+        self.next_phase += 1;
+        Ok(witness
+            .advice
+            .into_iter()
+            .enumerate()
+            .map(|(column_index, advice)| {
+                if column_indices.contains(&column_index) {
+                    Some(advice)
+                } else {
+                    None
+                }
+            })
+            .collect())
+    }
+}
+
+/// TODO: Document. Frontend function
+pub fn compile_circuit<F: Field, ConcreteCircuit: Circuit<F>>(
+    k: u32,
+    circuit: &ConcreteCircuit,
+    compress_selectors: bool,
+) -> Result<
+    (
+        CompiledCircuitV2<F>,
+        ConcreteCircuit::Config,
+        ConstraintSystem<F>,
+    ),
+    Error,
+> {
+    let n = 2usize.pow(k);
+    let mut cs = ConstraintSystem::default();
+    #[cfg(feature = "circuit-params")]
+    let config = ConcreteCircuit::configure_with_params(&mut cs, circuit.params());
+    #[cfg(not(feature = "circuit-params"))]
+    let config = ConcreteCircuit::configure(&mut cs);
+    let cs = cs;
+
+    if n < cs.minimum_rows() {
+        return Err(Error::not_enough_rows_available(k));
+    }
+
+    let mut assembly = crate::plonk::keygen::Assembly {
+        k,
+        fixed: vec![Polynomial::new_empty(n, F::ZERO.into()); cs.num_fixed_columns],
+        permutation: permutation::keygen::Assembly::new(n, &cs.permutation),
+        selectors: vec![vec![false; n]; cs.num_selectors],
+        usable_rows: 0..n - (cs.blinding_factors() + 1),
+        _marker: std::marker::PhantomData,
+    };
+
+    // Synthesize the circuit to obtain URS
+    ConcreteCircuit::FloorPlanner::synthesize(
```

Comment:
Can you not use acronyms in `pub` fn documentation? It's a bit confusing for newcommers.

Why do we have a hardcoded `FloorPlanner` here? Shouldn't we be able to choose here? Notice that `assembly` will depend on the layout done for copies in the permutation argument. Same can happen with selector compression.

Reply:
This is copied from original code:
https://github.com/privacy-scaling-explorations/halo2/blob/0c3e3b569519b86653a64f412bbce17e4e8acac4/halo2_proofs/src/plonk/keygen.rs#L235

I agree it's better to not use acronyms in pub documentation, but this is an internal comment.

The `FloorPlanner` here is an associated type, so it's not hardcoded.

Reply:
It's hardcoded in the sense that the user can't choose (isn't a generic left there that the user can define while calling the fn).

Reply:
Ah I see.  Yeah, the current design of the halo2 API is that the developer of the circuit chooses the floor planner, and not the user of the circuit.  I personally think this is OK; as the developer of the circuit may have insights on which floor planner is best for that particular circuit.  Also, there aren't many choices of `FloorPlanner`s :P

Reply:
Moved this discussion to https://github.com/privacy-scaling-explorations/halo2/issues/250